x/leverage/types/leverage.go

```diff
@@ -0,0 +1,137 @@
+package types
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// ComputeLiquidation takes the conditions preceding a liquidation and outputs the amounts
+// of base token that should be repaid, collateral uToken burned, and reward token allocated
+// as a result of the transaction, after accounting for limiting factors with as little
+// rounding as possible. Inputs are as follows:
+// - availableRepay: The lowest (in repay denom) of either liquidator balance, max repayment, or borrowed amount.
+// - availableCollateral: The amount of the reward uToken denom which borrower has as collateral
+// - availableReward: The amount of unreserved reward tokens in the module balance
+// - repayTokenPrice: The oracle price of the base repayment denom
+// - rewardTokenPrice: The oracle price of the base reward denom
+// - uTokenExchangeRate: The uToken exchange rate from collateral uToken denom to reward base denom
+// - liquidationIncentive: The liquidation incentive of the token reward denomination
+// - closeFactor: The dynamic close factor computed from the borrower's borrowed value and liquidation threshold
+// - borrowedValue: The borrower's borrowed value in USD
+func ComputeLiquidation(
+	availableRepay sdk.Int,
+	availableCollateral sdk.Int,
+	availableReward sdk.Int,
+	repayTokenPrice sdk.Dec,
+	rewardTokenPrice sdk.Dec,
+	uTokenExchangeRate sdk.Dec,
+	liquidationIncentive sdk.Dec,
+	closeFactor sdk.Dec,
+	borrowedValue sdk.Dec,
+) (tokenRepay sdk.Int, collateralBurn sdk.Int, tokenReward sdk.Int) {
```

Comment:
Refactor note: This function, which does not require a reference to the leverage keeper as it uses its inputs in isolation, has been move to leverage types for more efficient testing. (It doesn't require a test keeper or suite).

I plan to take a similar approach for other computations (#1168) in other PRs

Reply:
Moved back to original location - functionality improvement is retained.